src/client/render/pipeline.h

```diff
@@ -0,0 +1,391 @@
+	/**
+	 * Activate the render target and configure OpenGL state for the output.
+	 * This is usually done by @see RenderStep implementations.
+	 */
+	virtual void activate(PipelineContext *context)
+	{
+		m_clear = false;
+	}
+
+	/**
+	 * Resets the state of the object for the next pipeline iteration
+	 */
+	virtual void reset(PipelineContext *context) override
+	{
+		m_clear = true;
+	}
+
+protected:
+	bool m_clear {true};
+};
+
+/**
+ * Texture buffer represents a framebuffer with a multiple attached textures.
+ *
+ * @note Use of TextureBuffer requires use of gl_FragData[] in the shader
+ */
+class TextureBuffer : public RenderSource, public RenderTarget
+{
+public:
+	virtual ~TextureBuffer() override;
+
+	/**
+	 * Configure fixed-size texture for the specific index
+	 * 
+	 * @param index index of the texture
+	 * @param size width and height of the texture in pixels
+	 * @param height height of the texture in pixels
+	 * @param name unique name of the texture
+	 * @param format color format
+	 */
+	void setTexture(u8 index, core::dimension2du size, const std::string& name, video::ECOLOR_FORMAT format);
+
+	/**
+	 * Configure relative-size texture for the specific index
+	 * 
+	 * @param index index of the texture
+	 * @param scale_factor relation of the texture dimensions to the screen dimensions
+	 * @param name unique name of the texture
+	 * @param format color format
+	 */
+	void setTexture(u8 index, v2f scale_factor, const std::string& name, video::ECOLOR_FORMAT format);
+
+	/**
+	 * @Configure depth texture and assign index
+	 * 
+	 * @param index index to use for the depth texture
+	 * @param size width and height of the texture in pixels
+	 * @param name unique name for the texture
+	 * @param format color format
+	 */
+	void setDepthTexture(u8 index, core::dimension2du size, const std::string& name, video::ECOLOR_FORMAT format);
+
+	/**
+	 * @Configure depth texture and assign index
+	 * 
+	 * @param index index to use for the depth texture
+	 * @param scale_factor relation of the texture dimensions to the screen dimensions
+	 * @param name unique name for the texture
+	 * @param format color format
+	 */
+	void setDepthTexture(u8 index, v2f scale_factor, const std::string& name, video::ECOLOR_FORMAT format);
+
+	virtual u8 getTextureCount() override { return m_textures.size(); }
+	virtual video::ITexture *getTexture(u8 index) override;
+	virtual void activate(PipelineContext *context) override;
+	virtual void reset(PipelineContext *context) override;
+private:
+	struct TextureDefinition
+	{
+		bool valid { false };
+		bool fixed_size { false };
+		bool dirty { false };
+		v2f scale_factor;
+		core::dimension2du size;
+		std::string name;
+		video::ECOLOR_FORMAT format;
+	};
+
+	bool ensureTexture(video::ITexture **texture, const TextureDefinition& definition, PipelineContext *context);
+
+	video::IVideoDriver *m_driver { nullptr };
+	std::vector<TextureDefinition> m_definitions;
+	core::array<video::ITexture *> m_textures;
```

Comment:
is there a reason this is not a vector?

Reply:
Compatibility with video::IRenderTarget::setTexture(const core::array<> &, video::ITexture *).